tests/ui/meta/no_std-extern-libc.rs fails `aarch64-apple`

https://github.com/rust-lang/rust/pull/138192#issuecomment-2708197462

https://productionresultssa3.blob.core.windows.net/actions-results/e8a216b2-85ea-4389-b53a-e57abbb56654/workflow-job-run-1cf57105-ac8f-5932-9828-ef1ff1985697/logs/job/job-logs.txt?rsct=text%2Fplain&se=2025-03-08T10%3A56%3A57Z&sig=ZFpmubZ5r3pS%2BK%2BRuJIQ3iuz3vKQU9C1TYwgel36e%2Fs%3D&ske=2025-03-08T19%3A38%3A55Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-03-08T07%3A38%3A55Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-01-05&sp=r&spr=https&sr=b&st=2025-03-08T10%3A46%3A52Z&sv=2025-01-05

```

2025-03-08T10:44:43.6083130Z test [ui] tests/ui/windows-subsystem/console.rs ... ok
2025-03-08T10:44:43.6284930Z test [ui] tests/ui/writing-to-immutable-vec.rs ... ok
2025-03-08T10:44:43.7131990Z test [ui] tests/ui/zero-sized/zero-size-type-destructors.rs ... ok
2025-03-08T10:44:43.7743240Z test [ui] tests/ui/write-fmt-errors.rs ... ok
2025-03-08T10:44:43.8212490Z test [ui] tests/ui/zero-sized/zero-sized-binary-heap-push.rs ... ok
2025-03-08T10:44:43.9243640Z test [ui] tests/ui/zero-sized/zero-sized-linkedlist-push.rs ... ok
2025-03-08T10:44:43.9775230Z test [ui] tests/ui/zero-sized/zero-sized-btreemap-insert.rs ... ok
2025-03-08T10:44:44.0011260Z test [ui] tests/ui/zero-sized/zero-sized-tuple-struct.rs ... ok
2025-03-08T10:44:44.0592320Z test [ui] tests/ui/wrong-hashset-issue-42918.rs ... ok
2025-03-08T10:44:44.0593810Z 
2025-03-08T10:44:44.0594020Z failures:
2025-03-08T10:44:44.0594240Z 
2025-03-08T10:44:44.0594500Z ---- [ui] tests/ui/meta/no_std-extern-libc.rs stdout ----
2025-03-08T10:44:44.0594790Z 
2025-03-08T10:44:44.0596720Z error: test compilation failed although it shouldn't!
2025-03-08T10:44:44.0597030Z status: exit status: 1
2025-03-08T10:44:44.0600030Z command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" "/Users/runner/work/rust/rust/tests/ui/meta/no_std-extern-libc.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/work/rust/rust/vendor" "--sysroot" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2" "--target=aarch64-apple-darwin" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/meta/no_std-extern-libc" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers"
2025-03-08T10:44:44.0603270Z stdout: none
2025-03-08T10:44:44.0603600Z --- stderr -------------------------------
2025-03-08T10:44:44.0603920Z error[E0464]: multiple candidates for `rlib` dependency `libc` found
2025-03-08T10:44:44.0653450Z ##[error]  --> /Users/runner/work/rust/rust/tests/ui/meta/no_std-extern-libc.rs:8:1
2025-03-08T10:44:44.0660200Z    |
2025-03-08T10:44:44.0660370Z LL | extern crate libc;
2025-03-08T10:44:44.0660590Z    | ^^^^^^^^^^^^^^^^^^
2025-03-08T10:44:44.0660750Z    |
2025-03-08T10:44:44.0661250Z    = note: candidate #1: /Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib/liblibc-4564a71f990472d8.rlib
2025-03-08T10:44:44.0662130Z    = note: candidate #2: /Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib/liblibc-8545c79b49c06b96.rmeta
2025-03-08T10:44:44.0662610Z 
2025-03-08T10:44:44.0662850Z error: aborting due to 1 previous error
2025-03-08T10:44:44.0663040Z 
2025-03-08T10:44:44.0663210Z For more information about this error, try `rustc --explain E0464`.
2025-03-08T10:44:44.0663530Z ------------------------------------------
2025-03-08T10:44:44.0663700Z 
2025-03-08T10:44:44.0663700Z 
2025-03-08T10:44:44.0663710Z 
2025-03-08T10:44:44.0663780Z failures:
2025-03-08T10:44:44.0663980Z     [ui] tests/ui/meta/no_std-extern-libc.rs
2025-03-08T10:44:44.0664160Z 
2025-03-08T10:44:44.0664430Z test result: FAILED. 18393 passed; 1 failed; 313 ignored; 0 measured; 0 filtered out; finished in 390.62s
2025-03-08T10:44:44.0664750Z 
2025-03-08T10:44:44.0665000Z Some tests failed in compiletest suite=ui mode=ui host=aarch64-apple-darwin target=aarch64-apple-darwin
2025-03-08T10:44:44.0665430Z Build completed unsuccessfully in 0:10:24
```

--- a/src/bootstrap/src/core/build_steps/tool.rs
+++ b/src/bootstrap/src/core/build_steps/tool.rs
@@ -121,10 +121,18 @@ impl Step for ToolBuild {
 
         match self.mode {
             Mode::ToolRustc => {
-                builder.ensure(compile::Std::new(self.compiler, self.compiler.host));
-                builder.ensure(compile::Rustc::new(self.compiler, target));
+                // If compiler was forced, its artifacts should be prepared earlier.
+                if !self.compiler.is_forced_compiler() {
+                    builder.ensure(compile::Std::new(self.compiler, self.compiler.host));
+                    builder.ensure(compile::Rustc::new(self.compiler, target));
+                }
             }
-            Mode::ToolStd => builder.ensure(compile::Std::new(self.compiler, target)),
+            Mode::ToolStd => {
+                // If compiler was forced, its artifacts should be prepared earlier.
+                if !self.compiler.is_forced_compiler() {
+                    builder.ensure(compile::Std::new(self.compiler, target))
+                }
+            }
             Mode::ToolBootstrap => {} // uses downloaded stage0 compiler libs
             _ => panic!("unexpected Mode for tool build"),
         }

